analyzers/src/SonarAnalyzer.CSharp/Rules/LoggersShouldBeNamedForEnclosingType.cs

```diff
@@ -108,10 +119,16 @@ private static bool IsValidMethod(IMethodSymbol method)
             && method.TypeParameters.Length == 1;
     }
 
-    private static bool MatchesEnclosingType(SyntaxNode argument, TypeDeclarationSyntax typeSyntax, SemanticModel model) =>
+    private static bool MatchesEnclosingType(SyntaxNode argument, SyntaxNode enclosingNode, SemanticModel model) =>
         model.GetTypeInfo(argument).Type is { } argumentType
-        && model.GetDeclaredSymbol(typeSyntax).GetSymbolType() is { } enclosingType
-        && (enclosingType.Equals(argumentType) || argumentType.TypeKind is TypeKind.TypeParameter); // we do not want to raise on CreateLogger<T> if T is not concrete
+        && EnclosingTypeSymbol(model, enclosingNode) is { } enclosingType
+        && (enclosingType.Equals(argumentType)
+            || argumentType.TypeKind is TypeKind.TypeParameter  // Do not raise on CreateLogger<T> if T is not concrete
+            || enclosingType.DerivesOrImplementsAny(Loggers));  // Do not raise on Decorator pattern
+
+    private static ITypeSymbol EnclosingTypeSymbol(SemanticModel model, SyntaxNode enclosingNode) =>
+        model.GetDeclaredSymbol(enclosingNode).GetSymbolType()          // inside type
+        ?? model.GetSymbolInfo(enclosingNode).Symbol?.ContainingType;   // inside new(...)
```

Comment:
I think coverage is trolling me, both of these are hit.

Reply:
I think it's the question mark. Probably there is no case when the symbol cannot be found.

Reply:
It is equivalent to:
```csharp
(model.GetDeclaredSymbol(enclosingNode) ?? model.GetSymbolInfo(enclosingNode).Symbol).GetSymbolType();
```
that does not need the null check.

This usage is common enough that it could be extracted away from this class and reused.

Reply:
I applied your suggested refactoring, and removed the questionmark.

About extracting it, I am not sure. 
This works specifically with `TypeDeclarationSyntax` and `NewObjectCreationSyntax`, so I think it's very specific.